### ISSUE TITLE
Duplicate match case

### ISSUE THREAD

**Issue by JelleZijlstra** [astral-sh/ruff#18013]
### Summary

I'd like Ruff to warn on code like this:

```
def f(x: object):
    match x:
        case int():
            pass
        case int():
            pass
```

I was writing a match like this involving a lot of types, and I'm worried I'll duplicate a case, which would make the second case silently not execute.

A type checker could discover this problem somewhat naturally, but at the moment none of mypy, pyright, or ty detect this. (The algorithm is something like: start the type of the match subject as T. After every case matching to a type M, update it to `T & ~M`, taking into account previous narrowings. Within each match block, narrow it down to `T & M`, and warn if that intersection becomes Never.)

In a linter, the heuristic for this rule could be that the AST of a match case is identical to that of a previous case. You'd want to include any `if` guards in the match, but exclude `as` clauses, and distribute `|` patterns at the top level.

**Comment by AlexWaygood** [comment]
ty seems to be the type checker that gets closest here, FWIW! Unlike mypy, pyright and pyrefly, we correctly reveal `Never` for the second branch here:

```py
def f(x: object):
    match x:
        case int():
            reveal_type(x)  # int
        case int():
            reveal_type(x)  # Never
```

the only bit we don't do is this second part:

> warn if that intersection becomes Never

We'd probably want to do that as part of a general lint that warns about unintentionally unreachable blocks of code. But that requires some design work before we can implement it for ty, and it isn't an immediate priority for us. In the meantime, a lint rule like the one you suggest seems like a reasonable suggestion to me.
